### PR TITLE
[TESTING] clean and minor refactor to execution_graph code

### PR DESCRIPTION
**Pull request by aocsa**

### CODE REVIEW

**Comment by kharoc** [comment]
1 RAL

========================================================
FAILED TESTS
========================================================
Empty DataFrame
Columns: [index, TestId]
Index: []
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 286/286
FAILED: 0/286
CRASH: 0/286
TOTAL: 286

2 Worker's

==============================
Count without group by
==============================

=============== New query: TEST_01 Test DataType.PARQUET =================
==============================================================
11_170  |  MergeAggregate(group=[{}], EXPR$0=[COUNT()], EXPR$1=[COUNT($0)])
  10_171  |  DistributeAggregate(group=[{}], EXPR$0=[COUNT()], EXPR$1=[COUNT($0)])
    9_172  |  ComputeAggregate(group=[{}], EXPR$0=[COUNT()], EXPR$1=[COUNT($0)])
      13_173  |  BindableTableScan(table=[[main, nation]], projects=[[0]], aliases=[[n_nationkey]])
==============================================================
==============================================================
11_170  |  MergeAggregate(group=[{}], EXPR$0=[COUNT()], EXPR$1=[COUNT($0)])
  10_171  |  DistributeAggregate(group=[{}], EXPR$0=[COUNT()], EXPR$1=[COUNT($0)])
    9_172  |  ComputeAggregate(group=[{}], EXPR$0=[COUNT()], EXPR$1=[COUNT($0)])
      13_173  |  BindableTableScan(table=[[main, nation]], projects=[[0]], aliases=[[n_nationkey]])
==============================================================
111111111111111111111111111111111111
distributed.worker - WARNING -  Compute Failed
Function:  check_meta
args:      (Empty DataFrame
Columns: [, ll)n_nationkey)]
Index: [], Empty DataFrame
Columns: [(), ll)n_nationkey)]
Index: [], 'from_delayed')
kwargs:    {}
Exception: ValueError('Metadata mismatch found in `from_delayed`.\n\nPartition type: `cudf.core.dataframe.DataFrame`\n+--------+-------+----------+\n| Column | Found | Expected |\n+--------+-------+----------+\n|        | int64 | -        |\n| ()     | -     | int64    |\n+--------+-------+----------+')

Traceback (most recent call last):
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/ubuntu/.conda/envs/e2e/blazingdb-testing/BlazingSQLTest/EndToEndTests/allE2ETest.py", line 153, in <module>
    main()
  File "/home/ubuntu/.conda/envs/e2e/blazingdb-testing/BlazingSQLTest/EndToEndTests/allE2ETest.py", line 92, in main
    countWithoutGroupByTest.main(dask_client, drill, dir_data_file, bc, nRals)
  File "/home/ubuntu/.conda/envs/e2e/blazingdb-testing/BlazingSQLTest/EndToEndTests/countWithoutGroupByTest.py", line 58, in main
    executionTest()
  File "/home/ubuntu/.conda/envs/e2e/blazingdb-testing/BlazingSQLTest/EndToEndTests/countWithoutGroupByTest.py", line 36, in executionTest
    runTest.run_query(bc, drill, query, queryId, queryType, worder, '', acceptable_difference, use_percentage, fileSchemaType)
  File "/home/ubuntu/.conda/envs/e2e/blazingdb-testing/BlazingSQLTest/Runner/runTest.py", line 835, in run_query
    result_gdf = result_gdf.compute()
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/dask/base.py", line 166, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/dask/base.py", line 437, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/distributed/client.py", line 2595, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/distributed/client.py", line 1894, in gather
    asynchronous=asynchronous,
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/distributed/client.py", line 778, in sync
    self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/distributed/utils.py", line 348, in sync
    raise exc.with_traceback(tb)
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/distributed/utils.py", line 332, in f
    result[0] = yield future
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/distributed/client.py", line 1753, in _gather
    raise exception.with_traceback(traceback)
  File "/home/ubuntu/.conda/envs/e2e/lib/python3.7/site-packages/dask/dataframe/utils.py", line 665, in check_meta
    "%s" % ((" in `%s`" % funcname if funcname else ""), errmsg)
ValueError: Metadata mismatch found in `from_delayed`.

Partition type: `cudf.core.dataframe.DataFrame`
+--------+-------+----------+
| Column | Found | Expected |
+--------+-------+----------+
|        | int64 | -        |
| ()     | -     | int64    |
+--------+-------+----------+

File: engine/src/execution_graph/logic_controllers/BatchProcessing.h [hunk 1]
```diff
@@ -371,6 +360,50 @@ class Print : public PhysicalPlan {
 	std::ostream * ofs = nullptr;
 	std::mutex print_lock;
 };
+ 
+
+class OutputKernel : public kernel {
+public:
+	OutputKernel() : kernel("OutputKernel") {  }
+	virtual kstatus run() {
+		output = std::move(this->input_.get_cache()->pullFromCache());
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.h:369]: Why does this run only need to operate once? Is that becuase it is connected to some special cache that will concat all the subsets first?

File: engine/src/execution_graph/logic_controllers/BatchProcessing.h [hunk 2]
```diff
@@ -371,6 +360,50 @@ class Print : public PhysicalPlan {
...
+	frame_type	release() {
+		return std::move(output);
+	}
+
+protected:
+	frame_type output;
+};
+
+
+namespace test {
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.h:382]: can we move testing code to other files or was that arduous for some reason?

File: engine/src/execution_graph/logic_controllers/taskflow/graph.h [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	}
+	else if (config.type == CacheType::SIMPLE or config.type == CacheType::FOR_EACH) {
+		machine =  std::make_shared<ral::cache::CacheMachine>();
+	} else if (config.type == CacheType::CONCATENATING) {
+		machine =  std::make_shared<ral::cache::ConcatenatingCacheMachine>();
+	}
+	return machine;
+}
+
+static std::vector<std::shared_ptr<ral::cache::CacheMachine>> create_cache_machines(const cache_settings& config) {
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/taskflow/graph.h:23]: I am loathe to the cache machiens being put into a vector likek this. In all honesty I would much prefer a graph where the cache maciens and transform engines are actually connected to each other and both implement a "NODE" interface or somethign like that.

Comment (felipeblazing) [reply engine/src/execution_graph/logic_controllers/taskflow/graph.h:23]: That being said this is just a comment. mothing to do right now

File: engine/src/execution_graph/logic_controllers/taskflow/graph.h [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+			}
+		}
+	}
+
+	void execute() {
+		check_and_complete_work_flow();
+
+		std::vector<BlazingThread> threads;
+		std::set<std::pair<size_t, size_t>> visited;
+		std::deque<size_t> Q;
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/taskflow/graph.h:84]: what the fuck is Q? This is not a very appropriate name.

File: engine/src/execution_graph/logic_controllers/BatchProcessing.h (hunk 1 above)
Comment (aocsa) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.h:369]: Yes the OutputKernel uses a ConcatenatingCacheMachine.

File: engine/src/execution_graph/logic_controllers/BatchProcessing.h (hunk 2 above)
Comment (aocsa) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.h:382]: all right

File: engine/src/execution_graph/logic_controllers/taskflow/graph.h (hunk 4 above)
Comment (aocsa) [reply engine/src/execution_graph/logic_controllers/taskflow/graph.h:84]: all right, I will change in next PR.

**Pull request merged**